modules/loggers/mod_log_config.c

```diff
@@ -1602,13 +1634,204 @@ static ap_log_writer *ap_log_set_writer(ap_log_writer *handle)
     return old;
 }
 
-static apr_status_t ap_default_log_writer( request_rec *r,
+/**
+ * tries to convert the utf-8 encoded byte sequence pointed to by idx
+ * into a Unicode code point
+ *
+ * returns
+ *     number of bytes consumed, if valid utf-8 (1, 2, 3 or 4 bytes)
+ *     or  0 if idx is invalid
+ *     or -1 if byte sequence pointed to by idx is not valid utf-8
+ *
+ * the target code point cp is only updated in case of a valid utf-8 byte
+ * sequence
+ */
+static int utf8_next_code_point(const char* string, int idx, apr_uint32_t *cp)
+{
+    apr_uint32_t code_point = 0;
+    int len = 0;
+
+    int sl = strlen(string);
+    if(idx < 0 || idx >= sl)
+        return 0;
+
+    unsigned char c = string[idx];
+    int expect_more_bytes = 0;
+    if((c & 0b11110000) == 0b11110000) { // 4 byte utf-8
+        code_point = c & 0b00000111;
+        code_point <<= 3;
+        expect_more_bytes = 3;
+        len++;
+    } else if((c & 0b11100000) == 0b11100000) { // 3 byte utf-8
+        code_point = c & 0b00001111;
+        code_point <<= 4;
+        expect_more_bytes = 2;
+        len++;
+    } else if((c & 0b11000000) == 0b11000000) { // 2 byte utf-8
+        code_point = c & 0b00011111;
+        code_point <<= 5;
+        expect_more_bytes = 1;
+        len++;
+    } else if((c & 0b10000000) == 0b10000000) { // 1 byte utf-8 continuation
+        // this is an actual error, probably wrong idx
+        // idx did point to a byte which looks like an utf-8 1 byte continuation
+        return -1;
+    } else if(c < 0x80) { // 1 byte utf-8
+        code_point = c;
+        expect_more_bytes = 0;
+        len++;
+    }
+
+    while(expect_more_bytes > 0) {
+        if(++idx >= sl) break;
+        c = string[idx];
+        if((c & 0b10000000) == 0b10000000) { // 1 byte utf-8 continuation
+            c = c & 0b00111111;
+            code_point |= c;
+        } else {
+            // unexpected byte sequence, error
+            return -1;
+        }
+
+        expect_more_bytes--;
+        len++;
+
+        if(expect_more_bytes > 0) {
+            code_point <<= 6;
+        }
+    }
+
+    *cp = code_point;
+    return len;
+}
+
+/* see https://www.rfc-editor.org/rfc/rfc8259#section-7 */
+static int json_needs_encoding(const char* string)
+{
+    apr_uint32_t code_point;
+    int idx = 0;
+    int len = utf8_next_code_point(string, idx, &code_point);
+
+    while(len > 0) {
+        if(code_point  < 0x20 ||
+           code_point == 0x22 ||
+           code_point == 0x5c) {
+            return 1; // true, this json utf-8 string, needs encoding
+        }
```

Comment:
There could be a json_code_point_needs_encoding() helper since this same check is in json_encode() below too.

Reply:
Maybe `0x7f` (another control character) is missing?

Reply:
Strangley 0x7f is not part of the codepoints that needs escaping in JSON strings, see https://www.rfc-editor.org/rfc/rfc8259#page-9
Searching for JSON and 0x7f seems to at least reveal on bug report https://github.com/dominictarr/JSON.sh/issues/46

Reply:
I think `0x7f` is covered in https://www.rfc-editor.org/rfc/rfc8259#section-7
```
   All Unicode characters may be placed within the
   quotation marks, except for the characters that MUST be escaped:
   quotation mark, reverse solidus, and the control characters (U+0000
   through U+001F).
```

Reply:
Hm, scratch that, it stops at U+001F (not U+007F). TIL..